Data/HashMap/Internal.hs

```diff
@@ -218,19 +229,44 @@ instance NF.NFData2 Leaf where
 -- Invariant: The length of the 1st argument to 'Full' is
 -- 2^bitsPerSubkey
 
+
+-- these come from
+-- https://github.com/haskell-unordered-containers/hashable/blob/master/src/Data/Hashable/Class.hs#L221
+-- Can't have negatives in a natural.
+--
+-- to convert negatives use in ghci:
+-- >>> fromInteger (-3750763034362895579) :: Word64
+
+#if MIN_VERSION_hashable(1,3,1)
+-- fnv constant: https://github.com/haskell-unordered-containers/hashable/blame/c0cd4aa5238a2424d602719da9db69cc01bd17bb/src/Data/Hashable/Class.hs#L207
+type DefaultSalt = 14695981039346656037 -- fromInteger (-3750763034362895579) :: Word64
+#else
+type DefaultSalt = 15868100553162883236 -- old values https://github.com/haskell-unordered-containers/hashable/blame/ade7f97d1c59e9cfbad49b6ed130b90805311758/Data/Hashable/Class.hs#L202
+#endif
+
 -- | A map from keys to values.  A map cannot contain duplicate keys;
 -- each key can map to at most one value.
-data HashMap k v
+data HashMapT (salt :: Nat) k v
     = Empty
-    | BitmapIndexed !Bitmap !(A.Array (HashMap k v))
+    | BitmapIndexed !Bitmap !(A.Array (HashMapT salt k v))
     | Leaf !Hash !(Leaf k v)
-    | Full !(A.Array (HashMap k v))
+    | Full !(A.Array (HashMapT salt k v))
     | Collision !Hash !(A.Array (Leaf k v))
       deriving (Typeable)
 
-type role HashMap nominal representational
+-- backwards compatibility
+type HashMap = HashMapT DefaultSalt
 
-instance (NFData k, NFData v) => NFData (HashMap k v) where
+-- Allows coercion of a hashmap for example:
+--
+-- >>> newtype X = MkX Text
+-- >>> coerce (HashMap Text X) = HashMap Text Text
+--
+-- nominal role indicates that this isn't allowed.
+-- we don't want to coerce key's or salt, because it'd cause issues.
+type role HashMapT nominal nominal representational
```

Comment:
Somewhere, the documentation should point out [`sameNat`](https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-TypeNats.html#v:sameNat), which does a safe runtime check.

```haskell
sameNat :: (KnownNat a, KnownNat b) => proxy1 a -> proxy2 b -> Maybe (a :~: b)
```